client/client.go

```diff
@@ -300,17 +311,72 @@ func (client *Client) do(method, path string, query url.Values, headers map[stri
 		break
 	}
 	if err != nil {
-		return err
+		return nil, err
 	}
-	defer rsp.Body.Close()
+	return rsp, nil
+}
 
-	if v != nil {
-		if err := decodeInto(rsp.Body, v); err != nil {
-			return err
+// Do performs the HTTP request according to the provided options, possibly retrying GET requests
+// if appropriate for the status reported by the server.
+func (rq *defaultRequester) Do(ctx context.Context, opts *RequestOptions) (*RequestResponse, error) {
+	httpResp, err := rq.retry(ctx, opts.Method, opts.Path, opts.Query, opts.Headers, opts.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Is the result expecting a caller-managed raw body?
+	if opts.Type == RawRequest {
+		return &RequestResponse{Body: httpResp.Body}, nil
+	}
+
+	defer httpResp.Body.Close()
+	var serverResp response
+	if err := decodeInto(httpResp.Body, &serverResp); err != nil {
+		return nil, err
+	}
+
+	// Update the maintenance error state
+	if serverResp.Maintenance != nil {
```

Comment:
I know you didn't change this code, but I was just about to suggest "this is equivalent to just `rq.client.maintenance = serverResp.Maintenance`" but that's not true, because `serverResp.Maintenance` is a `*Error` and `rq.client.maintenance` is an `error`. [This whole thing.](https://go.dev/doc/faq#nil_error). Can we add a comment to explain why this is necessary to ensure this doesn't get "fixed" in future?

Reply:
```
+               // We cannot assign a nil pointer of type *Error to an
+               // interface here because the interface is only nil if
+               // both the type and value is nil.
+               // https://go.dev/doc/faq#nil_error
                rq.client.maintenance = nil
```

Reply:
Resolved.